metrics/storage/fio.sh

```diff
@@ -0,0 +1,296 @@
+#!/bin/bash
+#
+# Copyright (c) 2018 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# Description of the test:
+# Use fio to gather storate IO metrics.
+# The fio configuration can be modified via environment variables.
+# This test is only designed to handle a single file and class of job
+# in fio. If you require a more complex fio test then you are probably
+# better off writing that by hand or creating a new metrics test.
+
+set -e
+
+# General env
+SCRIPT_PATH=$(dirname "$(readlink -f "$0")")
+source "${SCRIPT_PATH}/../lib/common.bash"
+
+# Items for our local Dockerfile/image creation
+TEST_NAME="fio"
+IMAGE="local-fio"
+DOCKERFILE="${SCRIPT_PATH}/fio_dockerfile/Dockerfile"
+CONTAINER_NAME="fio_test"
+
+# How much RAM and how many (v)CPUs do we give the container under test?
+CONTAINER_RAM=${CONTAINER_RAM:-2G}
+CONTAINER_CPUS=${CONTAINER_CPUS:-1}
+
+# Important paths from the pov of both the host and the guest.
+# Some of them are mapped into both via a docker volume mount.
+HOST_OUTPUT_DIRNAME="${HOST_OUTPUT_DIRNAME:-${SCRIPT_PATH}/fio_output}"
+HOST_INPUT_DIRNAME="${HOST_INPUT_DIRNAME:-${SCRIPT_PATH}/fio_input}"
+# Used on the host side if we do a volume mount test.
+HOST_TEST_DIRNAME="${HOST_TEST_DIRNAME:-/tmp}"
+GUEST_OUTPUT_DIRNAME="/output"
+GUEST_INPUT_DIRNAME="/input"
+
+# This is the dir that fio will actually run the test upon in the container.
+# By default the tests will run on the container 'root mount'. If you set
+# TEST_VOLUME_MOUNT, then we will mount a volume mount over that test directory
+# and the tests will happen on a volume mount.
+TEST_VOLUME_MOUNT=${TEST_VOLUME_MOUNT:-0}
+GUEST_TEST_DIRNAME="/testdir"
+
+# These define which variety of the tests we will run
+#  Which of the read/write/rw tests will we run (we will run all
+#  that are set to 1).
+#
+# By default we do direct random read and write tests (not readwrite)
+FIO_READTEST=${FIO_READTEST:-1}
+FIO_WRITETEST=${FIO_WRITETEST:-1}
+FIO_READWRITETEST=${FIO_READWRITETEST:-0}
+FIO_DIRECT=${FIO_DIRECT:-1}
+FIO_RANDOM=${FIO_RANDOM:-1}
+
+# The blocksizes we will test. We have a separate set for direct or not
+# as direct mode can only use blocksize or larger sizes.
+FIO_BLOCKSIZES="${FIO_BLOCKSIZES:-128 256 512 1k 2k 4k 8k 16k 32k 64k}"
+FIO_DIRECT_BLOCKSIZES="${FIO_DIRECT_BLOCKSIZES:-4k 8k 16k 32k 64k}"
+
+# Other FIO parameters that are tweakable by setting in the environment.
+# Values from here are directly injected into the fio jobfiles used for
+# running the tests.
+FIO_NUMJOBS=${FIO_NUMJOBS:-4}
+#  By default run a time base test
+FIO_TIMEBASED=${FIO_TIMEBASED:-1}
+#  And 60s seems a good balance to get us repeatable numbers in not too long a time.
+FIO_RUNTIME=${FIO_RUNTIME:-60}
```

Comment:
Should we consider adding ramp_time?

Reply:
Can do. I know we have used it before in some `fio` testing.
I've added it in as an over-rideable, and set the default to 0s.
It's unclear to me just what it 'warms up' and in what circumstances - mostly I can see it would be caches, and mostly we are interested in un-cached setups and results.